File: src/main/java/shopping/member/adapter/in/api/MemberController.java
package shopping.member.adapter.in.api;

import jakarta.validation.Valid;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import shopping.auth.adapter.in.api.TokenResponse;
import shopping.auth.adapter.in.web.RefreshTokenCookieManager;
import shopping.auth.service.AuthTokens;
import shopping.member.service.MemberService;

@RestController
@RequestMapping("/api/members")
@RequiredArgsConstructor
public class MemberController {
    private final MemberService memberService;
    private final RefreshTokenCookieManager refreshTokenCookieManager;

    @PostMapping("/register")
    public ResponseEntity<TokenResponse> register(
            @Valid @RequestBody RegisterRequest request,
            HttpServletResponse response
    ) {
        AuthTokens tokens = memberService.register(request);
        return buildTokenResponse(response, tokens, HttpStatus.CREATED);
    }

    @PostMapping("/login")
    public ResponseEntity<TokenResponse> login(
            @Valid @RequestBody LoginRequest request,
            HttpServletResponse response
    ) {
        AuthTokens tokens = memberService.login(request);
        return buildTokenResponse(response, tokens, HttpStatus.OK);
    }

    private ResponseEntity<TokenResponse> buildTokenResponse(
            HttpServletResponse response,
            AuthTokens tokens,
            HttpStatus status
    ) {
        refreshTokenCookieManager.write(response, tokens.refreshToken());
        TokenResponse tokenResponse = TokenResponse.from(tokens);
        if (status == HttpStatus.OK) {
            return ResponseEntity.ok(tokenResponse);
        }
        return ResponseEntity.status(status).body(tokenResponse);
    }
}

File: src/main/kotlin/shopping/interfaces/api/member/MemberController.kt
package shopping.interfaces.api.member

import com.failsafe.interfaces.api.ApiResponse
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import shopping.application.member.MemberFacade

@RestController
@RequestMapping("/api/members")
class MemberController(
    private val memberFacade: MemberFacade,
) {
    @PostMapping("/register")
    fun register(
        @RequestBody request: MemberDto.RegisterRequest,
    ): ApiResponse<MemberDto.RegisterResponse> =
        memberFacade
            .register(request)
            .let { MemberDto.RegisterResponse.from(it) }
            .let { ApiResponse.success(it) }

    @PostMapping("/login")
    fun login(
        @RequestBody request: MemberDto.LoginRequest,
    ): ApiResponse<MemberDto.LoginResponse> =
        memberFacade
            .login(request)
            .let { MemberDto.LoginResponse.from(it) }
            .let { ApiResponse.success(it) }
}
